core/interop/Streams/ManagedLib/StreamAsIStream.cs

```diff
@@ -0,0 +1,167 @@
+using System.Runtime.InteropServices;
+
+#if NETFRAMEWORK
+using IStream = System.Runtime.InteropServices.ComTypes.IStream;
+using STATSTG = System.Runtime.InteropServices.ComTypes.STATSTG;
+#else
+using System.Runtime.InteropServices.Marshalling;
+using IStream = InteropWithStream.IStream;
+using STATSTG = InteropWithStream.STATSTG;
+#endif // !NETFRAMEWORK
+
+namespace InteropWithStream;
+
+/// <summary>
+/// Stream data type projected as IStream instance.
+/// </summary>
+public static unsafe partial class StreamExports
+{
+#if NETFRAMEWORK
+    [DNNE.Export]
+#else
+    [UnmanagedCallersOnly]
+#endif // !NETFRAMEWORK
+    public static int IStream_CreateStream(void** istream)
+    {
+        try
+        {
+            MemoryStream stream = new();
+#if NETFRAMEWORK
+            *istream = (void*)Marshal.GetIUnknownForObject(new MemoryStreamWrapper(stream));
+#else
+            *istream = ComInterfaceMarshaller<MemoryStreamWrapper>.ConvertToUnmanaged(new MemoryStreamWrapper(stream));
+#endif // !NETFRAMEWORK
+        }
+        catch (Exception e)
+        {
+            return e.HResult;
+        }
+        return 0;
+    }
+
+#if NETFRAMEWORK
+    [DNNE.Export]
+#else
+    [UnmanagedCallersOnly]
+#endif // !NETFRAMEWORK
+    public static int IStream_PrintStream(void* streamMaybe)
+    {
+        try
+        {
+#if NETFRAMEWORK
+            var streamWrapper = (MemoryStreamWrapper)Marshal.GetObjectForIUnknown((IntPtr)streamMaybe);
+#else
+            var streamWrapper = ComInterfaceMarshaller<MemoryStreamWrapper>.ConvertToManaged(streamMaybe);
+#endif // !NETFRAMEWORK
+            if (streamWrapper is null)
+            {
+                throw new NotSupportedException();
+            }
+
+            var stream = streamWrapper.Stream;
+            var data = stream.ToArray();
+            const int row = 4;
+            for (int i = 0; i < data.Length; ++i)
+            {
+                char delim = (i % row == row - 1) ? '\n' : ' ';
+                Console.Write($"{data[i]}{delim}");
+            }
+            Console.WriteLine();
+        }
+        catch (Exception e)
+        {
+            return e.HResult;
+        }
+        return 0;
+    }
+
+#if NETFRAMEWORK
+    [ComVisible(true)]
+    sealed partial class MemoryStreamWrapper : IStream
+    {
+        public MemoryStreamWrapper(MemoryStream s)
+        {
+            this.Stream = s;
+        }
+
+        public MemoryStream Stream { get; private set; }
+
+        // ISequentialStream
+        public void Read([MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 1), Out] byte[] pv, int cb, IntPtr pcbRead)
```

Comment:
Should this rather take `byte* pv`? As written, the unmanaged -> managed marshalling has to create an extra copy of the buffer as `byte[]` that is quite inefficient.

Reply:
Good suggestion. I can do that.